### ISSUE TITLE
Migrate from mysql 5.7 to 8.0 -- vacation table fails

### ISSUE THREAD

**Issue by mxyoung** [postfixadmin/postfixadmin#595]
I am attempting to migrate to a new server. Old server is running mysql 5.7, new server 8.0. I do a mysqldump on the 5.7 server, import that into the 8.0 server, and it fails with "Cannot add foreign key constraint."

I basically want to just recreate the vacation and vacation_notification tables (no problem with them being empty), but can't figure out exactly how. Is there a SQL file for the latest postfixadmin (3.3.10) for recreating tables?

**Comment by DavidGoodwin** [comment]
Does it give you a line number in the error message, to highlight where it's failed (and narrow down which constraint it might be complaining about).

**Comment by mxyoung** [comment]
Unfortunately I don't get a lot back and haven't been able to determine what constraint is problematic. Doing this from mysql command line:

mysql> CREATE TABLE `vacation` (
    ->   `email` varchar(255) NOT NULL,
    ->   `subject` varchar(255) NOT NULL,
    ->   `body` text NOT NULL,
    ->   `activefrom` timestamp NOT NULL DEFAULT '2000-01-01 06:00:00',
    ->   `activeuntil` timestamp NOT NULL DEFAULT '2038-01-18 06:00:00',
    ->   `cache` text NOT NULL,
    ->   `domain` varchar(255) NOT NULL,
    ->   `interval_time` int(11) NOT NULL DEFAULT '0',
    ->   `created` datetime NOT NULL DEFAULT '2000-01-01 00:00:00',
    ->   `modified` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
    ->   `active` tinyint(1) NOT NULL DEFAULT '1',
    ->   PRIMARY KEY (`email`),
    -> ) ENGINE=InnoDB DEFAULT CHARSET=latin1 COMMENT='Postfix Admin - Virtual Vacation';

The response is:  ERROR 1215 (HY000): Cannot add foreign key constraint

When I go to postfixadmin/upgrade.php?debug=1 I get back:

**CREATE TABLE IF NOT EXISTS `config` ( `id` int(11) not null auto_increment primary key, `name` VARCHAR(20) /*!40100 CHARACTER SET latin1 COLLATE latin1_general_ci */ NOT NULL DEFAULT '', `value` VARCHAR(20) /*!40100 CHARACTER SET latin1 COLLATE latin1_general_ci */ NOT NULL DEFAULT '', UNIQUE name ( `name` ) ) COMMENT = 'PostfixAdmin settings'

Database is up to date: 1844/0**

It doesn't seem to check for other tables or recreate them if they don't exist. When I run setup.php, I get no errors in the apache log, and setup reports:

**Everything seems fine... attempting to create/update database structure

Database is up to date: 1844/0**

I have $CONF['vacation'] = 'NO'; in the config. But I am on this path because when I go to list.php I get an error:

**[Wed Feb 16 08:54:08.878624 2022] [php7:notice] [pid 51353] [client ip-masked] Invalid query: SQLSTATE[42S02]: Base table or view not found: 1146 Table 'mail.vacation' doesn't exist caused by SELECT * FROM `vacation` WHERE email = ? AND active = ? , referer: https://mydomain.com/postfixadmin/list.php?table=domain**

**Comment by DavidGoodwin** [comment]
I don't know what's going on there - the 'create table vacation ...' isn't specifying any foreign key constraints.


I assume you already have a 'config' table with something in it - which is why the setup.php isn't creating anything else.

**Comment by DavidGoodwin** [comment]
What other tables exist in your database? Is there a vacation_notification table that is of type MyISAM ?

e.g.

```
CREATE TABLE `vacation_notification` (
  `on_vacation` varchar(255) CHARACTER SET latin1 COLLATE latin1_general_ci NOT NULL,
  `notified` varchar(255) CHARACTER SET latin1 COLLATE latin1_general_ci NOT NULL DEFAULT '',
  `notified_at` timestamp NOT NULL DEFAULT current_timestamp(),
  PRIMARY KEY (`on_vacation`,`notified`),
  CONSTRAINT `vacation_notification_pkey` FOREIGN KEY (`on_vacation`) REFERENCES `vacation` (`email`) ON DELETE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COMMENT='Postfix Admin - Virtual Vacation Notifications'
```
(You can use `ALTER TABLE vacation_notification ENGINE=InnoDB` to move it over)

**Comment by mxyoung** [comment]
```
+-----------------------+
| Tables_in_mail        |
+-----------------------+
| admin                 |
| alias                 |
| alias_domain          |
| config                |
| domain                |
| domain_admins         |
| fetchmail             |
| log                   |
| mailbox               |
| quota                 |
| quota2                |
| vacation_notification |
+-----------------------+
```

All are already InnoDB according to "show table status".
```
+-----------------------+--------+---------+------------+------+----------------+-------------+-----------------+--------------+-----------+----------------+-------------                            --------+---------------------+------------+-------------------+----------+----------------+------------------------------------------------+
| Name                  | Engine | Version | Row_format | Rows | Avg_row_length | Data_length | Max_data_length | Index_length | Data_free | Auto_increment | Create_time                                     | Update_time         | Check_time | Collation         | Checksum | Create_options | Comment                                        |
+-----------------------+--------+---------+------------+------+----------------+-------------+-----------------+--------------+-----------+----------------+-------------                            --------+---------------------+------------+-------------------+----------+----------------+------------------------------------------------+
| admin                 | InnoDB |      10 | Dynamic    |    3 |           5461 |       16384 |               0 |            0 |         0 |           NULL | 2022-02-15 1                            2:59:49 | 2022-02-15 12:59:50 | NULL       | latin1_swedish_ci |     NULL |                | Postfix Admin - Virtual Admins                 |
| alias                 | InnoDB |      10 | Dynamic    |  179 |            274 |       49152 |               0 |        16384 |         0 |           NULL | 2022-02-15 1                            2:59:51 | 2022-02-15 12:59:52 | NULL       | latin1_swedish_ci |     NULL |                | Postfix Admin - Virtual Aliases                |
| alias_domain          | InnoDB |      10 | Dynamic    |    0 |              0 |       16384 |               0 |        32768 |         0 |           NULL | 2022-02-15 1                            2:59:52 | NULL                | NULL       | latin1_swedish_ci |     NULL |                | Postfix Admin - Domain Aliases                 |
| config                | InnoDB |      10 | Dynamic    |    0 |              0 |       16384 |               0 |        16384 |         0 |              2 | 2022-02-15 1                            2:59:54 | 2022-02-15 12:59:55 | NULL       | latin1_swedish_ci |     NULL |                | PostfixAdmin settings                          |
| domain                | InnoDB |      10 | Dynamic    |   10 |           1638 |       16384 |               0 |            0 |         0 |           NULL | 2022-02-15 1                            2:59:55 | 2022-02-15 12:59:56 | NULL       | latin1_swedish_ci |     NULL |                | Postfix Admin - Virtual Domains                |
| domain_admins         | InnoDB |      10 | Dynamic    |    4 |           4096 |       16384 |               0 |        16384 |         0 |              5 | 2022-02-15 1                            2:59:56 | 2022-02-15 12:59:57 | NULL       | latin1_swedish_ci |     NULL |                | Postfix Admin - Domain Admins                  |
| fetchmail             | InnoDB |      10 | Dynamic    |    0 |              0 |       16384 |               0 |            0 |         0 |              1 | 2022-02-15 1                            2:59:58 | NULL                | NULL       | latin1_swedish_ci |     NULL |                |                                                |
| log                   | InnoDB |      10 | Dynamic    |  495 |            198 |       98304 |               0 |        32768 |         0 |            518 | 2022-02-15 1                            2:59:59 | 2022-02-15 13:00:00 | NULL       | latin1_swedish_ci |     NULL |                | Postfix Admin - Log                            |
| mailbox               | InnoDB |      10 | Dynamic    |  137 |            478 |       65536 |               0 |        16384 |         0 |           NULL | 2022-02-15 1                            3:00:01 | 2022-02-15 13:00:02 | NULL       | latin1_swedish_ci |     NULL |                | Postfix Admin - Virtual Mailboxes              |
| quota                 | InnoDB |      10 | Dynamic    |    0 |              0 |       16384 |               0 |            0 |         0 |           NULL | 2022-02-15 1                            3:00:02 | NULL                | NULL       | latin1_swedish_ci |     NULL |                |                                                |
| quota2                | InnoDB |      10 | Dynamic    |    0 |              0 |       16384 |               0 |            0 |         0 |           NULL | 2022-02-15 1                            3:00:03 | NULL                | NULL       | latin1_swedish_ci |     NULL |                |                                                |
| vacation_notification | InnoDB |      10 | Dynamic    |    0 |              0 |       16384 |               0 |            0 |         0 |           NULL | 2022-02-15 0                            9:14:02 | NULL                | NULL       | utf8_general_ci   |     NULL |                | Postfix Admin - Virtual Vacation Notifications |
+-----------------------+--------+---------+------------+------+----------------+-------------+-----------------+--------------+-----------+----------------+-------------                            --------+---------------------+------------+-------------------+----------+----------------+------------------------------------------------+

```


```
mysql> describe vacation_notification;

+-------------+--------------+------+-----+-------------------+-------------------+
| Field       | Type         | Null | Key | Default           | Extra             |
+-------------+--------------+------+-----+-------------------+-------------------+
| on_vacation | varchar(255) | NO   | PRI | NULL              |                   |
| notified    | varchar(255) | NO   | PRI |                   |                   |
| notified_at | timestamp    | NO   |     | CURRENT_TIMESTAMP | DEFAULT_GENERATED |
+-------------+--------------+------+-----+-------------------+-------------------+
3 rows in set (0.00 sec)

```

**Comment by mxyoung** [comment]
OK, so maybe this helps. I noticed the vacation_notification table was not using latin1_general_ci. I dropped that table, and now the create table for vacation works. But then I tried to re-create the vacation_notification table, and got this error:

`ERROR 3780 (HY000): Referencing column 'on_vacation' and referenced column 'email' in foreign key constraint 'vacation_notification_pkey' are incompatible.
`

**Comment by DavidGoodwin** [comment]
doing a "show create table vacation_notification" etc would give more detail and might highlight how the fields aren't compatible.

**Comment by mxyoung** [comment]
OK I figured this out. Providing details which may help someone else. First, the "show create table vacation_notification" doesn't show anything when the table create fails.

The incompatibility error was because the vacation_notification table was defining the vacation_notification.on_vacation field as "CHARACTER SET latin1", and the email field in the vacation table did not specify the character set. Even though the default character set of the vacation table is latin1, mysql 8.0 decided those fields those fields were incompatible because the data types did not match. I removed the CHARACTER SET definitions from the vacation_notification table, and both tables are now created and functioning. 

Note that this was not a problem in mysql 5.7. Perhaps there is a global setting somewhere that would help with this situation, but I don't know what it might be. I am also unclear what unintended consequences might arise from removing the CHARACTER SET specification on the fields in vacation_notify, though. Or if there is a better way to handle this.
